### ISSUE TITLE
Guides: Update TwoRavens documentation

### ISSUE THREAD

**Issue by pdurbin** [IQSS/dataverse#2212]
We will be previewing changes here (builds every hour from branch 4.0.1 currently):
- http://guides.dataverse.org/en/next/user/data-exploration/index.html
- http://guides.dataverse.org/en/next/installation/r-rapache-tworavens.html

@vjdorazio will push to this branch, which we will eventually merge: https://github.com/IQSS/dataverse/tree/2212-two-ravens

**Comment by pdurbin** [comment]
All the links above are 404s. The builds are gone, the branch is gone. If this update got off the ground I'm not aware of it.

@eaquigley can you please weigh in on this issue, especially with regard to the User Guide?

For the Installation Guide portion, I'm not planning on touching how http://guides.dataverse.org/en/4.2.3/installation/r-rapache-tworavens.html looks as part of the reorg for #2884 since that page was written by @landreev who is the authority. I'm just going to link to that page as an optional feature of Dataverse that an installation can choose to set up.

**Comment by pdurbin** [comment]
I mentioned to @landreev that I didn't touch any of the TwoRavens/Rserve/rApache stuff in my pull request at #2895. He agrees that section could use a rewrite at some point.

**Comment by pdurbin** [comment]
Related: pull request #3690.

**Comment by landreev** [comment]
OK, some of the ancient discussion in this issue is (likely) no longer relevant. 
But, since we've kept it open all this time - I'm resuscitating it, and I will use it for this new reiteration of the installer scripts and docs.

**Comment by landreev** [comment]
@dlmurphy 
Derek, 
You can look at the doc in question directly on github: 
https://github.com/IQSS/dataverse/blob/2212-newandimproved-tworavens-installer/doc/sphinx-guides/source/installation/r-rapache-tworavens.rst

if you do that, the formatting will be slightly different from what it looks like when we process our guides with swift, with our own styling; the colors and fonts are different, links look weird, etc. 
So, alternatively you can at the guide on my own dev. system here: 
http://140.247.116.186:8080/guides/installation/r-rapache-tworavens.html

**Comment by landreev** [comment]
@dlmurphy 

OK, improving documentation is an infinite task (so why bother doing it at all? - j/k)
I've already got caught today in fixing and rewriting more of it today...
But I feel that this particular guide is important; so we may want to invest an extra effort into making it readable and clear. It's not because it's going to be read by too many people. (It's really something that needs to be done once, for every Dataverse installation out there). But so far virtually everybody who has attempted to install this thing has written to us with questions and problems. TwoRavens has always been tricky to install; mostly because it needs a set of dependencies set up just the right way... And it's been tricky overall simply because we are trying to support a piece of software that somebody else wrote. (So, an important goal of the guide, among other things, to create the right level of expectations, and to communicate clearly that it's not our software really, that we'll try to help people installing it and integrating it with Dataverse; but for any help with the actual features they need to try and contact the actual developers...)

So, I really want your professional opinion on the whole thing... I've written good documentation, if I'm to say so myself... But it always takes me a very long time. Also, whenever I feel that there is something that's potentially confusing, I address it by explaining the living crap out of it, in painful detail. But that often makes it too verbose and tl;dr... So if you can suggest ways to make it communicate the same amount of information in fewer words, it'll be most appreciated. 

I definitely want this guide to be readable, and possible to follow. But I also feel fairly strongly that getting this thing to work does require some level of technical skills - with compiling code, installing Linux components and using R. So if, say, you don't feel you possess these skills and by the end (or the middle) of reading through this guide you end up feeling like it's not something you're even willing to attempt - then it could be considered the desired outcome too. As in, I want to communicate that one has to be at least a bit of a sysadmin to work on this task.

**Comment by dlmurphy** [comment]
Thanks for the update and the explanation, @landreev. It's really helpful to know where you're coming from when putting this guide together. I've started looking it over, and it is very clearly written so far. I see a few instances where I may be able to condense things a slight bit while preserving all details. With any changes I make, I'll be very careful not to cut any important information, especially since I don't fully understand some of the more technical stuff.

**Comment by dlmurphy** [comment]
My edits to the docs are all in!

**Comment by landreev** [comment]
Per conversation with @pdurbin : 
Added a new section (3.f.), explaining how to run a simple test of the newly installed TwoRavens, using the Dataverse-supplied dummy data file "50by1000"; with some screen shots and trouble shooting tips. 
(should it be part of the "Appendix" instead? - feedback welcome)

**Comment by landreev** [comment]
Per conversations with @pdurbin : 

items on the todo list:

- Instead of pointing them to the master distribution of TwoRavens, point by the specific commit - this way we'll insure that no later updates break the process as described. (will do this once everything else is complete; i.e., once I know that no further checkins into the TwoRavens code tree are necessary. Alternatively, may be cleaner to do this by tagging...)

- Change/rearrange the section 3.c. - specifically, do not instruct the user to go ahead and run the installer, before outlining some configuration choices they will be asked to make; specifically, on ports. This is currently offered in the reverse order...

**Comment by landreev** [comment]
Another tiny todo item: add the table of contents to the top of the guide.

**Comment by landreev** [comment]
Another todo: explain that the "disable directory listing in apache" is for security reasons.

**Comment by pdurbin** [comment]
As I mentioned at standup, this morning I successfully installed TwoRavens and friends for the first time on https://dev1.dataverse.org using the new instructions and script in Leonid's pull request. It's great work. I'm much more confident about the whole process.

The only step that I continue to hate is this one where you copy a file into directory of a deployed war file:

`cp /var/www/html/dataexplore/rook/preprocess/preprocess.R /usr/local/glassfish4/glassfish/domains/domain1/applications/dataverse-4.6.1/WEB-INF/classes/edu/harvard/iq/dataverse/rserve/scripts/preprocess.R`

It sounds like #3372 was opened to address this but the solution in there is somewhat fancy and I would favor a more simple approach to get that `preprocess.R` file out of the war file immediately and have `src/main/java/edu/harvard/iq/dataverse/rserve/RemoteDataFrameService.java` find it on disk via a configuration option.

There are other tweaks I would make to the doc but I'll wait for the dust to settle on that branch before I get in there and introduce merge conflicts.

Finally, I updated the diagram in the Installation Guide to explain where rApache is in the mix. Here's the before and after:

![screen shot 2017-03-21 at 12 39 10 pm](https://cloud.githubusercontent.com/assets/21006/24158726/7e5758fc-0e33-11e7-8deb-2c7b104085ef.png)

**Comment by pdurbin** [comment]
@landreev I mentioned this in passing yesterday but I think we should remove "you'll probably want to run Glassfish on ports 80 and 443" because we shouldn't be encouraging people to run Glassfish as root in order to be on ports 80 and 443.  In 24c6b4e we started saying, "Configure Glassfish to run as a user other than root" and someday for #1934 we should explain the specifics and provide an updated Glassfish init script.

On any server I set up I front Glassfish with Apache for a variety of reasons:

- I don't want to run Glassfish as root
- Putting an SSL cert in Glassfish is awful.
- Running Apache means I can configure Shibboleth if I want.

Basically, I think this new TwoRavens guide should recommend fronting with Apache.

Also, I added a diagram under "Overview". I hope it helps:

![tworavens_components](https://cloud.githubusercontent.com/assets/21006/24166112/a6a7773c-0e48-11e7-99aa-a8f8a80bcd98.png)

**Comment by pdurbin** [comment]
"@pdurbin just a note to say thank you for the updated diagram in #2212! when it comes time to refresh Odum's Dataverse bits there are a few things I'd like to do differently, and RApache is on my short list" -- @donsizemore at http://irclog.iq.harvard.edu/dataverse/2017-03-21

**Comment by landreev** [comment]
The last todo:s have been addressed: 

- download instructions modified to point at the specific checkin in TwoRavens github tree; 
- tworavens installer scipt section 3. rewritten, to offer the discussion of the configuration values first, before telling them to go ahead and run it...
- list of contents - added by @pdurbin;
- the "disable directory listing" part clarified by @pdurbin; 

finally, rewrote that "ports discussion" section of the appendix; that was the last part of this guide that i kinda hated... synchronizing it with what we're saying elsewhere in the doc - apache as the https front: recommended; https under glassfish: not recommended - made the section far more manageable and readable. 

@pdurbin , @dlmurphy  - could you please give the doc one more quick view, for typos and such. but it should have everything we wanted in terms of content by now.

**Comment by djbrooke** [comment]
We discussed this in the sprint planning meeting yesterday afternoon and this is where I'd usually put the notes, but the comment from @landreev did a better job of summarizing what's left/done. 

Thanks @landreev @pdurbin @dlmurphy for the efforts on this!

**Comment by djbrooke** [comment]
...and lest we forget, thanks @donsizemore for the encouragement !

**Comment by dlmurphy** [comment]
Made a couple tiny punctuation fixes. This is looking great to me.

**Comment by pdurbin** [comment]
Ok, I just did my final pass of the TwoRavens page (as of e75276f) and some final nitpicks that would be **optional** to change (mentioned to @landreev ):

- unbold "Rename the resulting directory"
- in the diagram, change Rserve from database icon to a box (my bad, sorry)
- swap the paragraphs in "overview" about "Estimate" and seeing the data pebbles. The pebbles come first in the "perform a quick test" steps later.
- when we talk about the hack we could link to #3372

Again, non of these are show stoppers for me. I'm fine with merging the pull request as-is.

**Comment by donsizemore** [comment]
Woo woo! So I took the rewritten installer documentation as an opportunity to break our test Dataverse's TwoRavens instance off onto its own host (Nevermore). Two things that might be noted in the current, public documentation, if not already in the working version:

• gcc-gfortran-4.8.5 is now included in the CentOS7 base repo
• RHEL/CentOS users will likely want mod_ssl alongside the httpd package

The mod_ssl note is boilerplate and very minor, but the gcc-gfortranm update could save an inexperienced user a bit of grief.

Happily TwoRavens-ing,
@donsizemore

**Comment by landreev** [comment]
OK, I think we are done with it. 
The latest cosmetic fixes are checked in. I say, merge it!

**Comment by landreev** [comment]
hmm. the formatting of the values table in the "run the installer" section got broken in the last checkin last night (so the table is not shown as of now). 
I'm going to go ahead and fix it quickly with a pull request that just adds 2 missing spaces there.

**Comment by landreev** [comment]
ok, done, sorry about that.

**Comment by landreev** [comment]
(by "done" I meant - made a pull request and merged it myself... it was literally 2 whitespaces. once again, sorry)
